### ISSUE TITLE
Masks not correspond to the RGB images

### ISSUE THREAD

**Issue by jiudawn** [CancerDataScience/CrowdsourcingDataset-Amgadetal2019#1]
Hello, thanks for your dataset. 
I downloaded all the RGB images and masks (with encoding 1,2,3...). But I found some of them are not correspond. The mask is like rotated but the RGB image not. like "TCGA-E2-A1L7-DX1_xmin51554_ymin15424_MPP-0.2300.png",  "TCGA-OL-A97C-DX1_xmin68058_ymin32495_MPP-0.2300.png" , their masks' display is tilted, but the RGB image is not. Why?

**Comment by kheffah** [comment]
@jiudawn Actually, this is just the impression you get when visualizing, but the RGB and mask correspond to each other. The black parts of the mask (zero pixels) correspond to an irrelevant part of the RGB. Everything (RGB and mask) are relative to the unrotated slide, and zero pixels are to be ignored, this also applied to the specific slides you mentioned.

**Comment by cooperlab** [comment]
Please see the comment in the readme file:
https://github.com/CancerDataScience/CrowdsourcingDataset-Amgadetal2019/blob/master/README.md

"Please be aware that zero pixels represent regions outside the region of interest (“don’t care” class) and should be assigned zero-weight during model training; they do NOT represent an “other” class."

**Comment by kheffah** [comment]
You can do something like:
``rgb[mask == 0] = 0`
to check that they correspond.

**Comment by jiudawn** [comment]
Thanks for your  quick reply. 
I knew that zero pixels represent regions outside the region of interest (“don’t care” class). But as for image "TCGA-OL-A97C-DX1_xmin68058_ymin32495_MPP-0.2300.png", size is 6770*7205. The mask (e.g. class 1) is tilted while the downloaded RGB image is covering the entire plane which seems to be resized.  So in this situation, the RGB image is just for visulization? If I want to use the data for traing a model, should I crop images from original wsis or change the masks? I am very interested to your elaborate work. I would appreciate your help.

**Comment by kheffah** [comment]
@jiudawn even for the image you mentioned, the mask and RGB correspond to each other. Just ignore the part of the RGB where the mask is zero completely (it's just for visualization). To use this for model training, you have to assign zero weight to these pixels.

**Comment by jiudawn** [comment]
Oh, I got it. But for this ignore area, if I put it into the model, it will learn and predict a mask, which I think will affect the final result?

**Comment by kheffah** [comment]
@jiudawn yes, exactly, so what we did in our paper was ignore these pixels from the loss calculation so they don't have an effect on the training

**Comment by jiudawn** [comment]
Oh, thanks again for your reply! 
I am now trying to reproduce your work, but I get bad results, especially for the class 3 and 5 (other)->(pixel dice about 0.5-0.6). So I checked the predict and compared it to the mask in the test set then found this problem. Could I ask for your code about how to solve it?  email: 201720111063@scut.mail.edu.cn
Thanks a million!

**Comment by kheffah** [comment]
@jiudawn ,

No worries, here are some notes and resources to get you started: 

1- Make sure to read through the supplement for detailed information in how we trained the model. We color normalized all images before inputting into the model. The reference image for Reinhard normalization is the RGB image TCGA-A2-A3XS-DX1_xmin21421_ymin37486_MPP-0.2000, which is part of the dataset, but here's a [direct link](https://drive.google.com/file/d/1q75LCrP4ghccC2E0npHT0CZDaug2prBv/view?usp=sharing) to download it for your convenience.

2- Software repository that you can use and build on. I used [this repository](https://github.com/kheffah/tensorflow_FCN8Workflow) (or some past version of it) to do the experiments. Feel free to modify to your use. If I remember correctly, you'll probably need to modify the Scripts/Run_Settings.py file, then run the Scripts/Run.py file in predict_unlabeled mode, making sure the run settings correctly point to the model weights folder, and use 5 classes with class labels [1, 2, 3, 4, 5]. It's been a while since I ran this code, so you may have to make some modifications, but it should get you started. If you find it easier to use your own codebase, but still would like something to give you a head start, you should probably checkout this repo, which I relied on for much of my own codebase. 

5- I strongly recommend that you group the classification classes during training, as we describe in the supplement of our paper. here's the relevant part, but I still recommend taking a look at the supplement:

> The 16-layer, FCN-8 variant of VGG fully-convolutional network was used in our experiments. The network was trained to map pixels into five region classes: tumor, stroma, inflammatory infiltration, necrosis and other. Regions that belong to rare classes were grouped with predominant were classes where appropriate, as follows: Grouped with “tumor”: angioinvasion, DCIS; Grouped with “inflammatory infiltrates”: lymphocytes, plasma cells, other immune infiltrates. Each ROI was divided into overlapping 800x800 pixel tiles, and tiles where over 90% of the area was composed of the “don’t care” class were ignored. The tiles were stored into .tfrecords files to be used in training. Slides from these institutes were not used in training the final segmentation model on the core set (i.e. were used as an unseen testing set to report accuracy): OL, LL, C8, BH, AR, A7 and A1. The amount of overlap used (and therefore, the amount of shift-augmentation) was inversely proportional to the size of the region of interest; this was done to ensure balanced representation of various histologic patterns regardless of ROI size. Crop augmentation to further increase robustness of training; a randomly-located 768x768 pixel image was cropped on the fly from each tile (after loading in memory) and was used in model training. The models were trained on 3 GPUs with a per-GPU batch size of 4 (total batch size of 12) using data parallelization. Adam optimizer was used with a learning rate of 1e-5. Weighted categorical cross-entropy loss was used to mitigate class imbalance, with the weight associated with each class determined by [...].

Thanks again for your interest in our work and dataset. I hope this answers your questions and gets you started without problems.

Cheers!

**Comment by jiudawn** [comment]
Great! Thanks for your guidance! That helps me a lot!
Here I found the other problem:  the "stroma" label of visualied data in the website is different from the downloaded mask. 
then I check again and guess it probably has problem in your website. https://goo.gl/cNM4EL.
Here is the link to TCGA-A7-A0CE-01Z-00-DX1:
https://demo.kitware.com/histomicstk/histomicstk#?image=5bbdee39e629140048d01ab1&bounds=56798%2C18492%2C69741%2C24411%2C0

**Comment by kheffah** [comment]
@jiudawn Glad to help. Actually, this is just an artifact of the visualization on the website, since it does not allow "holes" so holes in polygons are represented by another overlapping polygon. Don't worry, the masks are correct :)

**Comment by jiudawn** [comment]
Nice, I can put my mind at ease.
Another problem, I find that there are different magnifications, some images are 20x. I would like to know how you deal with it. I just adjusted 40x to 20x directly (0.5 scale). Is that right?

**Comment by kheffah** [comment]
@jiudawn Yes, that's correct.
